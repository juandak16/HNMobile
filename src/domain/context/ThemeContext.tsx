import {ReactNode, createContext, useContext, useEffect, useState} from 'react'
import {Theme, friendlyColors, hackColors} from 'src/data/local/theme'

interface ThemeContextProps {
  colors: Theme
  toggleDarkMode: () => void
}

const ThemeContext = createContext<ThemeContextProps | undefined>(undefined)

interface ThemeProviderProps {
  children: ReactNode
}

export const useTheme = (): ThemeContextProps => {
  const context = useContext(ThemeContext)
  if (!context) {
    throw new Error('useTheme must be used within a ThemeProvider')
  }
  return context
}

export const ThemeProvider: React.FC<ThemeProviderProps> = ({children}) => {
  const [isHackMode, setIsHackMode] = useState(false)
  const [colors, setColors] = useState(friendlyColors)

  useEffect(() => {
    if (isHackMode) {
      setColors(hackColors)
    } else {
      setColors(friendlyColors)
    }
  }, [isHackMode])

  const toggleDarkMode = (): void => {
    setIsHackMode(prev => !prev)
  }

  const theme: ThemeContextProps = {
    colors,
    toggleDarkMode,
  }

  return <ThemeContext.Provider value={theme}>{children}</ThemeContext.Provider>
}
