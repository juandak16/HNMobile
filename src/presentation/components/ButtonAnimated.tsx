import {Animated, ViewStyle, StyleSheet} from 'react-native'
import React from 'react'
import {RectButton} from 'react-native-gesture-handler'
import {useTheme} from 'src/domain/context/ThemeContext'

type ButtonProps = {
  onPress: () => void
  dragX: Animated.AnimatedInterpolation<number | string>
  progress?: Animated.AnimatedInterpolation<number | string>
  children: React.ReactNode
  direction: 'left' | 'right'
}

export default function ButtonAnimated({onPress, dragX, progress, children, direction}: ButtonProps) {
  const {colors} = useTheme()
  const styles = StyleSheet.create({
    button: {
      flex: 1,
      padding: 15,
      width: '100%',
      backgroundColor: colors.buttonBackground,
      alignItems: 'center',
      justifyContent: 'center',
    },
    text: {
      padding: 0,
      width: '100%',
      color: colors.buttonText,
      fontSize: 12,
    },
  })
  const opacityStyle: Animated.WithAnimatedValue<ViewStyle> = {
    opacity: progress ? progress : 1,
  }

  const trans = dragX.interpolate({
    inputRange: direction === 'left' ? [-100, 0] : [0, 100],
    outputRange: [0, 0],
  })

  const transformStyle: Animated.WithAnimatedValue<ViewStyle> = {
    transform: [{translateX: trans}],
  }

  return (
    <Animated.View style={[transformStyle, opacityStyle]}>
      <RectButton onPress={onPress} style={styles.button}>
        <Animated.Text style={styles.text}>{children}</Animated.Text>
      </RectButton>
    </Animated.View>
  )
}
