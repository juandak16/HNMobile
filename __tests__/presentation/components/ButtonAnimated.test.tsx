import React from 'react'
import {render, fireEvent} from '@testing-library/react-native'
import {Animated} from 'react-native'
import {RectButton} from 'react-native-gesture-handler'
import {useTheme} from 'src/domain/context/ThemeContext'
import ButtonAnimated from 'src/presentation/components/ButtonAnimated'

// Mock dependencias
jest.mock('src/domain/context/ThemeContext')
jest.mock('react-native-gesture-handler', () => ({
  RectButton: jest.fn(({children, onPress}) => <button onClick={onPress}>{children}</button>),
}))

describe('ButtonAnimated', () => {
  beforeEach(() => {
    // Configura mocks según sea necesario
    ;(useTheme as jest.Mock).mockReturnValue({
      colors: {
        buttonBackground: '#ed5c53',
        buttonText: '#FFFFFF',
      },
    })
  })

  it('renders correctly', () => {
    const onPressMock = jest.fn()
    const dragXMock = new Animated.Value(0)

    const {getByText, toJSON} = render(
      <ButtonAnimated onPress={onPressMock} dragX={dragXMock} direction="left">
        Test Button
      </ButtonAnimated>,
    )

    expect(getByText('Test Button')).toBeTruthy()
    expect(toJSON()).toMatchSnapshot()
  })

  it('calls onPress function on button press', () => {
    const onPressMock = jest.fn()
    const dragXMock = new Animated.Value(0)

    const {getByText} = render(
      <ButtonAnimated onPress={onPressMock} dragX={dragXMock} direction="left">
        Test Button
      </ButtonAnimated>,
    )

    // Simula una pulsación en el botón
    fireEvent.press(getByText('Test Button'))

    // Verifica que la función onPressMock haya sido llamada
    expect(onPressMock).toHaveBeenCalled()
  })
})
