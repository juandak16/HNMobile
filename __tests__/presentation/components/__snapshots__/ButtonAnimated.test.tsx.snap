// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonAnimated renders correctly 1`] = `
<View
  collapsable={false}
  style={
    {
      "opacity": 1,
      "transform": [
        {
          "translateX": 0,
        },
      ],
    }
  }
>
  <button
    onClick={[MockFunction]}
  >
    <Text
      collapsable={false}
      style={
        {
          "color": "#FFFFFF",
          "fontSize": 12,
          "padding": 0,
          "width": "100%",
        }
      }
    >
      Test Button
    </Text>
  </button>
</View>
`;
